crates/chia-consensus/src/puzzle_fingerprint.rs

```diff
@@ -0,0 +1,412 @@
+        }
+        let buf = a.atom(arg);
+
+        // every atom gets a length prefix, to avoid playing games with the
+        // resulting hash.
+        // e.g. two adjacent atoms whose concatenation stays the same, but sizes
+        // changes. Those cases must be distinguished
+        fingerprint.update((buf.as_ref().len() as u32).to_be_bytes());
+        fingerprint.update(buf.as_ref());
+    }
+    Ok(args)
+}
+
+/// This functions runs a *trusted*, *dedup* puzzles, i.e. one that has already
+/// been fully / validated in mempool mode, and returns the cost and the
+/// conditions / fingerprint for it. The conditions fingerprint is a hash of its
+/// known / condition outputs. This is used for identical spend deduplication to
+/// compare / whether two spends are identical and can be deduplicated. The
+/// condition / fingerprint should not be expected to be stable across different
+/// chia_rs versions.
+/// This function will fail if the puzzle returns a condition that's not
+/// supported by DEDUP spends.
+pub fn compute_puzzle_fingerprint(
+    puzzle: &Program,
+    solution: &Program,
+    max_cost: Cost,
+    flags: u32,
+) -> core::result::Result<(Cost, [u8; 32]), ValidationErr> {
+    let flags = flags | MEMPOOL_MODE;
+
+    let mut a = Allocator::new_limited(500_000_000);
+    let (mut cost, conditions) = puzzle.run(&mut a, flags, max_cost, solution)?;
+    let mut iter = conditions;
+    let mut free_condition_countdown: usize = FREE_CONDITIONS;
+
+    let mut fingerprint = Sha256::new();
+
+    while let Some((c, next)) = a.next(iter) {
+        iter = next;
+
+        let Some(op) = parse_opcode(&a, first(&a, c)?, flags) else {
+            // we just ignore unknown conditions
+            continue;
+        };
+
+        if (flags & COST_CONDITIONS) != 0 {
+            if free_condition_countdown == 0 {
+                cost += GENERIC_CONDITION_COST;
+            } else {
+                free_condition_countdown -= 1;
+            }
+        }
+
+        // since we only run in mempool mode, we don't need to take unknown
+        // conditions into account, including the ones with cost. This puzzle is
+        // expected to have already passed mempool-mode validation
+        match op {
+            AGG_SIG_UNSAFE
+            | AGG_SIG_ME
+            | AGG_SIG_PUZZLE
+            | AGG_SIG_PUZZLE_AMOUNT
+            | AGG_SIG_PARENT
+            | AGG_SIG_AMOUNT
+            | AGG_SIG_PARENT_PUZZLE
+            | AGG_SIG_PARENT_AMOUNT
+            | SEND_MESSAGE
+            | RECEIVE_MESSAGE => {
+                return Err(ValidationErr(NodePtr::NIL, ErrorCode::InvalidCondition));
+            }
+            CREATE_COIN => {
+                cost += CREATE_COIN_COST;
+
+                // CREATE_COIN, puzzle_hash, amount
+                let rest = hash_atom_list(&mut fingerprint, &a, c, 3)?;
+
+                // make sure to include the hint if present. If it's not present
```

Comment:
We should probably include all memos in the hash, since memos can be important to applications, right? Or is this a DoS vector?

Reply:
how many memos are "all" of them? If it's infinite, then I think it's a DoS vector.
how deep is a memo tree allowed to be? If it's infinite, then I think it's also a DoS vector

Reply:
I think that sounds reasonable, at least as a first version